### PR TITLE
Adding case statement and passing context to the recursive call

### PR DESCRIPTION
**Pull request by Schmavery**
Weird, I could have sworn I submitted a PR of this a couple days ago but now I can't find it.

This addresses #16 by adding support for switch statements.  I added some capability to print it nicer if there are no glob patterns in the cases, but couldn't figure out how to traverse the AST.

The reason that there are a lot of changes is I tweaked the printer to pass forward a context object.  If you don't like this we can talk.

### CODE REVIEW

File: src/semantics.js [hunk 1]
```diff
@@ -160,73 +173,107 @@ var semicolonCmdNames = [
...
 
     inFunctionBody = true;
-    var blockString = block.toJS(this.args.indent);
+    var blockString = block.toJS(this.args.indent, this.args.ctx);
     inFunctionBody = false;
 
     return 'function ' + idStr + '(..._$args) ' + blockString;
   },
+  CaseCommand: function(_case, expr, _in, _ws, cases, _ws2, _esac){
+    var varName = "tmp" + Date.now();
```
Comment (nfischer) [inline src/semantics.js]: Would it be possible to change this to something more consistent? It makes testing a lot easier. If you do something like prefix some name with a `$`, then it should be safe to use in the translation (`$` can't be in a bash identifier, but it works for JS).

File: src/semantics.js [hunk 2]
```diff
@@ -160,73 +173,107 @@ var semicolonCmdNames = [
...
+    var indent = this.args.indent;
+
+    // If we can scan the children to see if there are glob patterns
+    // ahead of time, we can set hasGlob to false and generate cleaner code.
+    var hasGlob = true;
+
+    if (hasGlob){
```
Comment (nfischer) [inline src/semantics.js]: Does this do anything? Won't this always evaluate to true?

File: src/semantics.js [hunk 3]
```diff
@@ -160,73 +173,107 @@ var semicolonCmdNames = [
...
+      cases.toJS(indent + 1, this.args.ctx).join(nl(indent + 1)) +
+      nl(indent) + "}";
+    }
+  },
+  CaseCase: function(opts, _par, _ws, cmds, _ws2, _semisemi, comment){
+    var varName = this.args.ctx.caseVar;
+    var comment = comment.toJS(this.args.indent, this.args.ctx);
+    return opts.toJS(0, this.args.ctx)
+        .map(s => s.substr(1, s.length - 2))
+        .map(s => isGlobString(s) ? ("case " + s + ":")
```
Comment (nfischer) [inline src/semantics.js]: Could you rewrite this as `function () { }`? I'd like to keep compatibility with iOS safari 9.3 for a bit longer (at least until I upgrade my phone to iOS 10 :laughing:). Not a big concern, I can fix it up after the merge myself.

Comment (Schmavery) [inline src/semantics.js]: Sure, was just a habit.

File: src/semantics.js (hunk 2 above)
Comment (nfischer) [inline src/semantics.js]: Let's take it out for now (simplifies things), and we can add it back in later to prettify the output.

File: src/semantics.js (hunk 1 above)
Comment (nfischer) [inline src/semantics.js]: SGTM

**Comment by nfischer** [comment]
Saw the updates. I'll review when I get a chance and ping this thread for any changes. Thanks!

**Comment by nfischer** [comment]
@Schmavery Sorry about the delay. Tried this out, this looks awesome! Thanks for the contribution

Fixes #16

**Comment by nfischer** [comment]
Whoops, accidentally closed instead of merging. My mistake

**Pull request merged**